go/vt/vttablet/grpctmclient/cached_client.go

```diff
@@ -0,0 +1,340 @@
+		default:
+			dialer.m.Lock()
+			if conn, ok := dialer.conns[addr]; ok {
+				// Someone else dialed this addr while we were polling. No need
+				// to evict anyone else, just reuse the existing conn.
+				defer dialer.m.Unlock()
+				return dialer.redial(conn)
+			}
+
+			dialer.qMu.Lock()
+			conn := dialer.queue[0]
+			if conn.refs != 0 {
+				dialer.qMu.Unlock()
+				dialer.m.Unlock()
+				continue
+			}
+
+			// We're going to return from this point
+			defer dialer.m.Unlock()
+			defer dialer.qMu.Unlock()
+			heap.Pop(&dialer.queue)
+			delete(dialer.conns, conn.key)
+			conn.cc.Close()
+
+			return dialer.newdial(addr, false /* manage queue lock */)
+		}
+	}
+}
+
+// newdial creates a new cached connection, and updates the cache and eviction
+// queue accordingly. This must be called only while holding the write lock on
+// dialer.m as well as after having successfully acquired the dialer.connWaitSema. If newdial fails to create the underlying
+// gRPC connection, it will make a call to Release the connWaitSema for other
+// newdial calls.
+//
+// It returns the three-tuple of client-interface, closer, and error that the
+// main dial func returns.
+func (dialer *cachedConnDialer) newdial(addr string, manageQueueLock bool) (tabletmanagerservicepb.TabletManagerClient, io.Closer, error) {
+	dialerStats.ConnNew.Add(1)
+
+	opt, err := grpcclient.SecureDialOption(*cert, *key, *ca, *name)
+	if err != nil {
+		dialer.connWaitSema.Release()
+		return nil, nil, err
+	}
+
+	cc, err := grpcclient.Dial(addr, grpcclient.FailFast(false), opt)
+	if err != nil {
+		dialer.connWaitSema.Release()
+		return nil, nil, err
+	}
+
+	// In the case where dial is evicting a connection from the cache, we
+	// already have a lock on the eviction queue. Conversely, in the case where
+	// we are able to create a new connection without evicting (because the
+	// cache is not yet full), we don't have the queue lock yet.
+	if manageQueueLock {
+		dialer.qMu.Lock()
+		defer dialer.qMu.Unlock()
+	}
+
+	conn := &cachedConn{
+		TabletManagerClient: tabletmanagerservicepb.NewTabletManagerClient(cc),
+		cc:                  cc,
+		lastAccessTime:      time.Now(),
+		refs:                1,
+		index:               -1, // gets set by call to Push
+		key:                 addr,
+	}
+	heap.Push(&dialer.queue, conn)
+	dialer.conns[addr] = conn
+
+	return dialer.connWithCloser(conn)
+}
+
+// redial takes an already-dialed connection in the cache does all the work of
+// lending that connection out to one more caller. this should only ever be
+// called while holding at least the RLock on dialer.m (but the write lock is
+// fine too), to prevent the connection from getting evicted out from under us.
+//
+// It returns the three-tuple of client-interface, closer, and error that the
+// main dial func returns.
+func (dialer *cachedConnDialer) redial(conn *cachedConn) (tabletmanagerservicepb.TabletManagerClient, io.Closer, error) {
+	dialerStats.ConnReuse.Add(1)
```

Comment:
tioli, but `Inc()` does the same as `Add(1)`

Reply:
unfortunately we're using our own home-rolled version of stats, so `Add` is all i've got: https://github.com/vitessio/vitess/blob/main/go/stats/counters.go